## Initial data preparation

In [2]:
#imports
import numpy as np
import pandas as pd
from pandas.io.json import json_normalize
import json

import pprint
pp = pprint.PrettyPrinter()

# Business

In [37]:
business = [] #List of businesses in JSON format

filepath = "data/Business_Toronto_Restaurant.json"
with open(filepath) as fp:  
    line = fp.readline()
    while line:
        business.append(json.loads(line))
        line = fp.readline()

In [38]:
print(len(business))
pp.pprint(business[5])

7148
{'address': '8 Glen Watford Drive',
 'attributes': {'Ambience': {},
                'BestNights': {},
                'BusinessAcceptsCreditCards': False,
                'BusinessParking': {},
                'DietaryRestrictions': {},
                'GoodForKids': True,
                'GoodForMeal': {},
                'HairSpecializesIn': {},
                'Music': {},
                'RestaurantsReservations': False,
                'RestaurantsTakeOut': True},
 'business_id': 'dTWfATVrBfKj7Vdn0qWVWg',
 'categories': ['Restaurants', 'Chinese', 'Food Court'],
 'city': 'Toronto',
 'hours': {'Friday': '11:00-21:00',
           'Monday': '11:00-21:00',
           'Saturday': '11:00-21:00',
           'Sunday': '11:00-21:00',
           'Thursday': '11:00-21:00',
           'Wednesday': '11:00-21:00'},
 'is_open': 1,
 'latitude': 43.78706,
 'longitude': -79.27617,
 'name': 'Flavor Cuisine',
 'neighborhood': 'Scarborough',
 'postal_code': 'M1S 2C1',
 'review_count': 6,
 'stars':

In [39]:
business_df = json_normalize(business)
columns = []
for c in business_df.columns:
    columns.append("".join(c.split('attributes.')))
business_df.columns = columns

In [ ]:
pd.set_option('display.max_columns', 500)
print(len(business_df))
print(list(business_df.columns))
business_df.to_csv("data/business.csv",index=False)
business_df.head()

Dataset consists of information on 7148 restaurants in Toronto, Canada. The consists of a list of JSON objects. Each JSON object contains information such as the resuraunts name, whether the resturant is currently open and the resturants stars. A resturant stars is a value between 0 and 5. Before the data can be used it is necessary to transfor it <br><br>

There are also 155300 reviews these contain. This data set contained individual star ratings and text reviews. They can be linked to the business through the business ID. 

[Snap shot] <br>
Used the pandas.io.json import json_normalize function to transform the json data into a pandas dataframe. A pandas dataframe is a commpn data structure used in pyhton. It is similar to a sql database or csv table. Each row represents a resturant and the columns are the characteristics of that resurant. Subsequent preparation of this data base is required to answer the 3 questions. The columns of the dataset represented each of the attributes from the JSON data. 

# Old Reviews

In [31]:
reviews_old = [] #List of Reviews in JSON format

filepath = "data/Review_Toronto_Restaurant.json"
with open(filepath) as fp:  
    line = fp.readline()
    while line:
        reviews_old.append(json.loads(line))
        line = fp.readline()

In [32]:
print(len(reviews_old))
pp.pprint(reviews_old[5])

155300
{'business_id': 'Aov96CM4FZAXeZvKtsStdA',
 'cool': 1,
 'date': '2010-07-16',
 'funny': 0,
 'review_id': '23eqwlZzCWZkADWfd9atZw',
 'stars': 5,
 'text': 'This place is awesome! Definitely authentic!!!\n'
         '\n'
         'My two favourite dishes are the rice flour rolls and the chicken '
         "pho. The rice flour rolls are always fresh whenever I'm there! And "
         'the chicken pho is always flavourful!! mmmm....just thinking of it '
         'makes me want some!\n'
         '\n'
         'My boyfriend is Vietnamese and he agrees that this place is '
         'authentic and one of the best Vietnamese restaurants he has ever '
         'eaten at.\n'
         '\n'
         'Prices are very reasonable too!',
 'useful': 2,
 'user_id': 'u0LXt3Uea_GidxRW1xcsfg'}


In [34]:
reviews_old_df = json_normalize(reviews_old)

In [36]:
reviews_old_df['business_id'].nunique()

7051

In [42]:

print(sum(business_df['review_count']),len(reviews_old_df))

303757 155300


There appears to be some missing reviews in "Review_Toronto_Restaurant.json" There are 7148 resturant and therefore 7148 unique resturant IDs in the business dataset. However, if we count the unique IDs in the json file we only get 7051. In other words, reviews for 97 resturants are missing. 
<br>
Furthermore, if we sum the review_count column of the business dataset we get the total number of 303757. This is total number of individual reviews for resturants in Toronto. There should therefore be the same number of idividual reviews in the JSON file. However, there are only 155300 individual reviews. This suggests that nearly half of the reivews are missing. 

In [35]:
print(len(reviews_old_df))
reviews_old_df.head()

155300


business_id  cool        date  funny               review_id  \
0  9_CGhHMz8698M9-PkVf0CQ     2  2012-05-11      0  ymAUG8DZfQcFTBSOiaNN4w   
1  5r6-G9C4YLbC7Ziz57l3rQ     0  2013-02-09      0  w41ZS9shepfO3uEyhXEWuQ   
2  z8oIoCT1cXz7gZP5GeU5OA     0  2013-05-01      0  PIsUSmvaUWB00qv5KTF1xA   
3  XWTPNfskXoUL-Lf32wSk0Q     1  2011-09-28      0  PdZ_uFjbbkjtm3SCY_KrZw   
4  RtUvSWO_UZ8V3Wpj0n077w     1  2012-12-03      1  lsoSqIrrDbQvWpMvsSj2xw   

   stars                                               text  useful  \
0      4  Who would have guess that you would be able to...       0   
1      3  Not bad!! Love that there is a gluten-free, ve...       1   
2      4  This is currently my parents new favourite res...       1   
3      3  Server was a little rude.\n\nOrdered the calam...       5   
4      3  Wanted to check out this place due to all the ...       2   

                  user_id  
0  u0LXt3Uea_GidxRW1xcsfg  
1  u0LXt3Uea_GidxRW1xcsfg  
2  u0LXt3Uea_GidxRW1xcsfg  
3  u0LXt3Uea_GidxRW1xcsfg  
4  u0LXt3Uea_GidxRW1xcsfg

Talk about this one first + give screen shot to illustrate how data was transformed. The reviews dataset was read in and transformed to a csv file using the json_normalize function. 

# Obtain new reviews 

In [43]:
303780- 303757

23

It was therefore nessery to extract the reviews from the large review.json file. This was done by obtaining the list of business_id values from the business dataset. Then the review.json file was cycled through line by line (ie. each line is one review). The 'business_id' value for each review was checked and the review was only kept if the ID was in the list of business_ids. This collection of JSON reviews was then normalized line the Business dataset. A snapshot of this dataset can be seen if Figure X. This process resulted in a dataset 303780 reviews with 7148 unique ID's. Reviews for all the busines are now present. The total number of reivews is 23 more than the business review count. This suggests some reivews are not accounted for in the business dataset. 

<br>
Further analysis was conducted on the review datase and business dataset

In [11]:
IDs = list(business_df['business_id'])
len(IDs)

7148

In [23]:
reviews = [] #List of Reviews in JSON format
# 5261669 lines in total
count = 0
filepath = "data/review.json"
with open(filepath) as fp:  
    line = fp.readline()
    while line:
        review = json.loads(line)
        if review['business_id'] in IDs:
            reviews.append(review)
        line = fp.readline()
        if count%200000 == 0: 
            print("{}/5261669 procesed".format(count))
        count = count +1
        
reviews[0]

0/5261669 procesed
200000/5261669 procesed
400000/5261669 procesed
600000/5261669 procesed
800000/5261669 procesed
1000000/5261669 procesed
1200000/5261669 procesed
1400000/5261669 procesed
1600000/5261669 procesed
1800000/5261669 procesed
2000000/5261669 procesed
2200000/5261669 procesed
2400000/5261669 procesed
2600000/5261669 procesed
2800000/5261669 procesed
3000000/5261669 procesed
3200000/5261669 procesed
3400000/5261669 procesed
3600000/5261669 procesed
3800000/5261669 procesed
4000000/5261669 procesed
4200000/5261669 procesed
4400000/5261669 procesed
4600000/5261669 procesed
4800000/5261669 procesed
5000000/5261669 procesed
5200000/5261669 procesed


{'review_id': 'ymAUG8DZfQcFTBSOiaNN4w',
 'user_id': 'u0LXt3Uea_GidxRW1xcsfg',
 'business_id': '9_CGhHMz8698M9-PkVf0CQ',
 'stars': 4,
 'date': '2012-05-11',
 'text': "Who would have guess that you would be able to get fairly decent Vietnamese restaurant in East York? \n\nNot quite the same as Chinatown in terms of pricing (slightly higher) but definitely one of the better Vietnamese restaurants outside of the neighbourhood. When I don't have time to go to Chinatown, this is the next best thing as it is down the street from me.\n\nSo far the only items I have tried are the phos (beef, chicken & vegetarian) - and they have not disappointed me! Especially the chicken pho.\n\nNext time I go back, I'm going to try the banh cuon (steamed rice noodle) and the vermicelli!",
 'useful': 0,
 'funny': 0,
 'cool': 2}

In [24]:
check_IDs = []
for r in reviews:
    check_IDs.append(r['business_id'])

for c in check_IDs:
    if c not in IDs:
        print(c)

In [25]:
reviews_df = json_normalize(reviews)

In [ ]:
303757 

In [26]:
print(len(reviews_df))
reviews_df.to_csv("data/reviews.csv",index=False)
reviews_df.head()

303780


business_id  cool        date  funny               review_id  \
0  9_CGhHMz8698M9-PkVf0CQ     2  2012-05-11      0  ymAUG8DZfQcFTBSOiaNN4w   
1  5r6-G9C4YLbC7Ziz57l3rQ     0  2013-02-09      0  w41ZS9shepfO3uEyhXEWuQ   
2  z8oIoCT1cXz7gZP5GeU5OA     0  2013-05-01      0  PIsUSmvaUWB00qv5KTF1xA   
3  XWTPNfskXoUL-Lf32wSk0Q     1  2011-09-28      0  PdZ_uFjbbkjtm3SCY_KrZw   
4  RtUvSWO_UZ8V3Wpj0n077w     1  2012-12-03      1  lsoSqIrrDbQvWpMvsSj2xw   

   stars                                               text  useful  \
0      4  Who would have guess that you would be able to...       0   
1      3  Not bad!! Love that there is a gluten-free, ve...       1   
2      4  This is currently my parents new favourite res...       1   
3      3  Server was a little rude.\n\nOrdered the calam...       5   
4      3  Wanted to check out this place due to all the ...       2   

                  user_id  
0  u0LXt3Uea_GidxRW1xcsfg  
1  u0LXt3Uea_GidxRW1xcsfg  
2  u0LXt3Uea_GidxRW1xcsfg  
3  u0LXt3Uea_GidxRW1xcsfg  
4  u0LXt3Uea_GidxRW1xcsfg

In [28]:
reviews_df['business_id'].nunique()

7148